src/mg/PAMI/Client/Impl/ClientImpl.php

```diff
@@ -405,20 +405,21 @@ public function send(OutgoingMessage $message)
 	    if (@fwrite($this->_socket, $messageToSend) < $length) {
     	    throw new ClientException('Could not send message');
 	    }
-	    $read = 0;
-	    while($read <= $this->_rTimeout) {
-	        $this->process();
-	        $response = $this->getRelated($message);
-	        if ($response != false) {
-	            $this->_lastActionId = false;
-	            return $response;
-	        }
-	        usleep(1000); // 1ms delay
-	        if ($this->_rTimeout > 0) {
-	            $read++;
-	        }
-	    }
-	    throw new ClientException('Read timeout');
+	    while (1) {
+	    	stream_set_timeout($this->_socket, $this->_rTimeout ? $this->_rTimeout : 1);
```

Comment:
The while loop is require to pull in more data. Because we are running in non-blocking mode we might be receiving just part of a message. 

The while loop should actually monitor / check the connection timeout, which it currently not doing. Now that i think about it. 
If connectiontimeout is a maximum time we will observe asterisk then that would be ok. If it is meant as a wait time for any event to happen then we need to reset a counter if some data has come in. So it depends on your definiton of connectiontimeout.

I do agree that any error from stream_set_timeout should be handled.